src/main/java/org/openrewrite/staticanalysis/AddStaticModifierToPublicFinalConstantsAndFields.java

```diff
@@ -0,0 +1,94 @@
+ * limitations under the License.
+ */
+package org.openrewrite.staticanalysis;
+
+import org.openrewrite.ExecutionContext;
+import org.openrewrite.Recipe;
+import org.openrewrite.Tree;
+import org.openrewrite.java.JavaIsoVisitor;
+import org.openrewrite.java.tree.J;
+import org.openrewrite.java.tree.J.Modifier;
+import org.openrewrite.java.tree.Space;
+import org.openrewrite.marker.Markers;
+
+import java.time.Duration;
+import java.util.Collections;
+import java.util.Set;
+
+import static java.util.Collections.emptyList;
+
+public class AddStaticModifierToPublicFinalConstantsAndFields extends Recipe {
+
+    @Override
+    public String getDisplayName() {
+        return "Add `static` to `public final` variables";
+    }
+
+    @Override
+    public String getDescription() {
+        return "Finds fields declared as `public final` and adds `static` modifier, meanwhile sort all the modifiers.";
+    }
+
+    @Override
+    public Set<String> getTags() {
+        return Collections.singleton("RSPEC-S1170");
+    }
+
+    @Override
+    public Duration getEstimatedEffortPerOccurrence() {
+        return Duration.ofMinutes(3);
+    }
+
+    @Override
+    public JavaIsoVisitor<ExecutionContext> getVisitor() {
+        return new JavaIsoVisitor<ExecutionContext>() {
+
+            @Override
+            public J.ClassDeclaration visitClassDeclaration(J.ClassDeclaration classDecl, ExecutionContext ctx) {
+                J.ClassDeclaration enclosing = getCursor().getParent().firstEnclosing(J.ClassDeclaration.class);
+                if (enclosing == null || classDecl.hasModifier(Modifier.Type.Static)) {
+                    return super.visitClassDeclaration(classDecl, ctx);
+                }
+                // Ignore non-static inner classes as they can't have static fields
+                return classDecl;
+            }
+
+            @Override
+            public J.NewClass visitNewClass(J.NewClass newClass, ExecutionContext ctx) {
+                // Ignore anonymous inner classes as they can't have static fields
+                return newClass;
+            }
+
+            @Override
+            public J.VariableDeclarations visitVariableDeclarations(J.VariableDeclarations multiVariable,
+                                                                    ExecutionContext ctx) {
+                J.VariableDeclarations v = super.visitVariableDeclarations(multiVariable, ctx);
+                if (multiVariable.hasModifier(Modifier.Type.Public) &&
+                    multiVariable.hasModifier(Modifier.Type.Final) &&
+                    !multiVariable.hasModifier(Modifier.Type.Static)) {
+
+                    multiVariable.getModifiers().add(new J.Modifier(Tree.randomId(),
+                            Space.format(" "), Markers.EMPTY, " ",
+                            Modifier.Type.Static, emptyList()));
+
+                    return v.withModifiers(ModifierOrder.sortModifiers(multiVariable.getModifiers()));
+
+                }
```

Comment:
This might be too simplistic. As seen on for instance [this class](https://github.com/openrewrite/rewrite/pull/4610/files#diff-99f8ee8cca47323ff5db87e9429b0553dc7327cbe132b7bb1b66a58c0cf1ca27R84-R87), making the instance fields static could drastically change behavior if there's more than one instance, if combined with mutable fields. This operation should only be safe for immutable types like primitives.